src/main/java/javaslang/collection/HashSet.java

```diff
@@ -380,7 +390,7 @@ public boolean isEmpty() {
         for (T element : elements) {
             trie = trie.remove(element);
         }
-        return new HashSet<>(trie);
+        return trie.size() == list.get().length() ? this : new HashSet<>(trie);
```

Comment:
also here. if true.remove(element) returns the same instance, we may test with == instead of size and length